SEP Proposal: Extend ManageData for arbitrary on-chain data storage (#1090)
stellar/stellar-protocol: ecosystem/sep-0039.md

### NFT Issuance
Given an arbitrary digital asset, here's how issuance typically works:

  1. An account is created to represent the digital asset on the ledger
  2. The account stores the entire digital asset within its managed data entries.
  3. The account issues a Stellar asset representing ownership over the digital asset we just stored.
  4. Any account holding that Stellar asset lays claim to the digital asset.

Notice that ownership of the token directly conveys ownership of a unit of the asset, entirely within the Stellar ledger. Other scenarios such as locking down the issuing account to avoid supply increases, ownership transfers via payments, etc. are built on top of this basic model and are thus extraneous to this SEP.

Note also that this SEP does not address concerns around *immutability*. While there is an optional checksum component (addressed later), the account itself is free to modify the data at-will if it is not locked.

Our exclusive focus in this SEP is Step 2: **storing the asset within the account.**

### Data Format
We begin by describing the raw data format, then dive into each component.

>     00|version|metadata length|media-type;param1=value;param2=value;etc<binary data>

Reviewer: ```suggestion
    000|version|metadata length|media-type;param1=value;param2=value;etc<binary data>
```

Reviewer: ```suggestion
    initial index (000)|version|metadata length|media-type;param1=value;param2=value;etc<binary data>
```

>     00|version|metadata length|media-type;param1=value;param2=value;etc<binary data>

Reviewer: Are we printing this in raw string format vs encoding it along with the rest of the `<binary data>`? Seems like we should just concatenate the entire thing as a singular buffer. And only have the index maybe be the `radix of 36 index|version|metadata length|<metadata binary><fields binary>`. Honestly I think the entire sep39 format should be one super buffer with no consideration for any raw string format. I could be convinced the indexing could be the radix of 36, that's kinda clever and helps both sorting and avoiding clashes with non sep39 key:value pairs.

Reviewer: ```suggestion
    00<1 byte version buffer><2 byte metadata length buffer><metadata buffer><concatenated buffer of field values>
    e.g. 00|1|30|<text/plain;l=10;k=title,text/plain;l=50;k=description,image/png;l=500;k=image>|<concatenated buffer of field values>
```

Author: I was thinking more about the radix thing, and honestly _all_ we need from it is sortability from Horizon. It offers no other value. The decoder can ignore it entirely. With that in mind, we can even use `aa`, `ab`, ... through `zz`.

Reviewer: Let's stick with numbers. 00-9z is less likely to clash with a non sep39 key than strictly alphanumeric chars

Here, `|` is not a character, but rather a semantic divider used to separate the components. The following sections break down each component in detail.

> #### Index

Reviewer: I'm confused by this. Are you suggesting printing out the index raw via `000`, `001` etc? This seems inefficient. In my estimation this should definitely be converted to using two bytes. The way I've been doing it via JS is 
```
const index = Buffer.alloc(2)
index.writeIntBE({index}, 0, 2)
```

Then later on to actually sort the keys you can do
```
parseInt(key.slice(0, 2).toString('hex'), 16)
```

The method of using three chars burns one extra char per manage data entry.

Was the thinking that by using raw string indexes you could avoid collision with other non sep39 key value entries?

---------

**Author's note**: This portion is definitely up for debate. Base36 (well, more accurately, a radix of 36) allows us to shave a byte off of each row (saving up to 1kB total), but we could stick to a 000-999 model (stringified integer) for simplicity. Note that supporting it is trivial in any language, however, since we know it's strictly an integer. For example,

```python
import string 
ALPHABET = string.digits + string.ascii_lowercase

def int_to_b36(i: int) -> str:
    d, r = divmod(i, len(ALPHABET))
    return ALPHABET[d] + ALPHABET[r]

def b36_to_int(s: str) -> int:
    return ALPHABET.index(s[0]) * len(ALPHABET) + ALPHABET.index(s[1])
```

Another consideration is perhaps trying to have **two** indices within this value. Since we need 10 bits, we can use the other 6 for identifying this asset within the account itself. There's still the question of printables, but  though, so this needs more thought.

---------

#### Version
The first entry's index is followed by the version string. For the current version of this SEP, this is just `1`.

#### Metadata

> The **metadata** is a [Media Type](https://www.iana.org/assignments/media-types/media-types.xhtml) alongside any associated parameters. It is prefixed by a **length**, which is just a stringified integer (that is, a 32-byte metadata will have "32" as its length). When parsing, you should consume characters until hitting a non-digit, since a valid Media Type will never start with a digit. Note that row indices are not included in the metadata length.

Reviewer: Where is this length in this example? https://github.com/stellar/stellar-protocol/pull/1090/files#diff-f2073a8bc6d9084cf6715b7120504e920b4a8663547a7ee6065b2074573c4438R58

A prefix seems like it would clash with the initial metadata length.

Reviewer: Or are you saying there could be multiple metadata fields?

Or are you missing the length parameter? How do I know how long each media-type is in the actual `<binary data>` in your example?

Author: I think we covered this [below](https://github.com/stellar/stellar-protocol/pull/1090#discussion_r760711781)? Some miscommunication about lengths. Re-open if not :+1:

**Author's note**: If we go with a radix of 36, we could use that here, too.

There are no *required* parameters.

> Standard *optional* parameters include:

Reviewer: Ah I think I see you're keeping the <binary data> limited to _just_ image data vs title, description and image data broken up by using the `len` like we had previously. I don't like this. At least not if the idea is that this is a raw string. Descriptions can be quite long so this at _least_ needs to be encoded into the basE91 at the very least. I still think I like the binary concatenation with the `len` option to allow other buffers to be encoded into a super buffer. This would also allow support for multiple non-utf8 assets to be encoded into a single sep39 account.

Author: Going to respond to [this](https://github.com/stellar/stellar-protocol/pull/1090#discussion_r760707194) here as they're related, I think.

> Descriptions can be quite long so this at least needs to be encoded into the basE91 at the very least.

As I was defining the metadata format, I had trouble envisioning a world in which you have non-printable metadata. In that case, why wouldn't it just be part of the header? That is, what's wrong with `image/png;n=A title;d=This is the description.`? Why do they need to be encoded? A long description is still just printable ASCII; encoding it would just be unnecessary blowup. I guess you can't include Media-Type identifiers, like `;` or `=` in your description... but that's a bit of a weak argument.

Maybe it's a little wonky if you start getting cute with emojis or something, but the beauty of the metadata format is that at that point, you're free to do what you want, right? Anyone can define the multi-buffer interface on top of this themselves. For example, have the metadata be `image/png;n=Title;d=12` and just take the first 12 bytes of the returned binary and treat them like your description. _You_ have a specific use case, but does it generalize? Do _we_ need to impose specifics?

To be clear, I do see value in standardizing the multi-buffer approach with lengths in the meta, but I'm still not convinced that it's necessary when someone could define it "on top" of the existing format. Heck, even having a title and description be "official" optional parameters seems unnecessary to me. If you want them, include them.

Author: Ahhhh okay, sorry, I went back and read our earlier back-and-forth and I see the disconnect now. To me, there's was single media type and metadata (title, desc) went with it (the fact that it was `text/plain` went :arrow_right: over my head). Hence my suggestion to just use strings, define the layer yourself, etc.

You're saying: **let there be many media types and let the header define them.**

New suggestion, then (if we're sure we want to do this):

`media-type/subtype;s=<size>,media-type-2/subtype;s=<size>`

Size is optional if there's one media type, required otherwise. 

Drop name, description, etc. as optional parameters. Let that be part of the buffer itself because there's no reason for us to impose a particular convention about what a field is "called" (leave that up to the user In other words, if you have a title field, let it just say `text/plain;s=19` in the header, then it as `Title=Hello, world!` (or `title:Hello, world!` or just `Hello, world!` because you implicitly know it's the title) within the binary data itself. 

Point is, we don't need to enforce that in the SEP. Just require a size for multi-part.

Author: I think I address this in 383b47c.

Reviewer: > Let that be part of the buffer itself

Field titles absolutely should be part of the metadata. If they're baked into the buffer itself I now have to decode the buffer, split off an `=` or `:` then re-buffer to deliver to the client. The data should be only data so I can just deliver that buffer in a REST response. I agree we don't need to enforce any naming conventions. Just maybe have `n` for `name` be a suggested optional value. So we'd have `l` for length and `n` for name and `c` for checksum. If you opt to store the data in the metadata itself you use the `d` key.

Reviewer: `l` and `d` then would be mutually exclusive. Either, or or neither in the case of one or the last value in the metadata set of `meta-type`s

Author: My rationale was that "what if I don't care about a name, and I already know that the "text/plain" part of the buffer is the name?" But if it's optional, that concern is alleviated anyway.

Regarding `d`, I say we disallow that entirely. Let's require it to be in the buffer; it keeps parsing simple.

Reviewer: A strong reason to keep `d` is that for text/plain fields keeping it as plaintext vs converting to basE91 will keep things more efficient. And again in the case of a long description it could cross quite a few keys and thus forcing that to go through the basE91 encoding will lose some bytes.

Reviewer: On second thought any `=` `,` or `;` in a metadata `d` fields would decimate the ability to appropriately parse the metadata 🤔  yeah we'd better keep all data, plaintext or otherwise outside the metadata.

Author: Good rationale, but there's unfortunately a huge blocker: Media Types technically can't have whitespace :sweat_smile:. I didn't realize this earlier.

Reviewer: It's settled then. All data will live in the main binary data buffer with splitting denoted by the `l` key. `l` or `s`? I've seen both.

Author: Let's actually use `s` because it's easier to distinguish visually than `l` vs `1`.

  * `n`: this defines the name or "title" of your asset
  * `d`: this provides a description your asset

>   * `c`: a crc32 checksum of the asset

Reviewer: I've never heard of `crc32`. Could we get a little context here like you have for basE91?

Author: Sure, yeah. I'll start here: crc32 is a very common (definitely not esoteric like BasE91) checksum algorithm (see [Wikipedia](https://en.wikipedia.org/wiki/Cyclic_redundancy_check)) It's not designed with cryptographic principles (integrity, collision resistance, inversion, etc.) in mind like the SHA family is, but rather just with the goal of error-checking. It's a fast, simple way to check that you have the right data.

Metadata _must_ only be composed of printable ASCII characters.

#### Binary Data
With the header out of the way, we need to describe the way the remaining binary data should be represented in `ManageData` entries.

Since the _key_ is restricted specifically to printable characters (ref: Stellar Core's [`isString32Valid()`](https://github.com/stellar/stellar-core/blob/b031954458df3bb31d8d98f136c6fee40523a10d/src/util/types.cpp#L89-L101) and C's [`isprint()`](https://en.cppreference.com/w/cpp/string/byte/isprint)), this gives us an encoding space of 95 characters or ~6.5 bits per character (a ~25% reduction in the space). From our research, the [BasE91] encoding scheme is closest to this theoretical limit.

The _value_ can be any binary data; thus, no encoding is necessary.

All of the binary data is thus prepended with an index, the next BasE91-encoded chunk sized to fit into the 64-length key string, and the next 64-byte binary chunk as the value.

If the final entry fits into the key (that is, the value is not necessary), the encoder _must_ set the value to an empty buffer (as opposed to `nil` or `null`). Otherwise, this will delete the entry.

> #### Implementation Note

Reviewer: I'm confused here. What is this solving for. If we can submit empty buffers as values we just keep chunking along until we run out of actual chunks and we finish with an empty buffer value.

Author: That's exactly what it's saying here: if you're out of data on the last row for the value, you still need set it to be an empty buffer. It's exactly what you said [here](https://github.com/stellar/stellar-protocol/pull/1090#discussion_r759331013).

Reviewer: Sorry I meant the `Implementation Note`
> Because you cannot predict the length of your encoded string without doing the encoding itself, it will be necessary to use trial-and-error to find the longest binary chunk that fits in <= 64 bytes. While this does mildly impact encoding time, it ensures optimality on-chain.

Author: Oh, this is actually related to your [later comment](https://github.com/stellar/stellar-protocol/pull/1090#issuecomment-984858571): base64 will always have 48 -> 64 because of padding, but base91 is less predictable (part of the reason why it's more efficient). That means you need to "guess and check" until you max out the buffer.

Author: It's impossible for base91 to do better than 10% overhead, so if you search from `ceil(n/1.10)` (where `n` is the final length you _want_), you'll get it. This was also defined in `_encode_nearest()` in the original commit (see [here](https://github.com/stellar/stellar-protocol/pull/1090/commits/9668dbe582319e0e83c5f427e028dee454411a43#diff-9e64d12ca10dd7b12e3632c74b968e5d35f418587ee64f6d92fb0d9aac347c65R136-R147)), if you want a reference implementation.

## Preamble

```
SEP: 0039

> Title: On-Chain Storage of Non-Fungible Assets

Reviewer 2: A little while ago @lydiat highlighted the benefits in having succinct SEP names.

Following the pattern of SEP-11 Txrep about:
```suggestion
Title: On-Chain NFTs: Storage of Non-Fungible Assets On-Chain
```

Reviewer: I'm of the opinion we should keep this sep laser focused on NFTs. If some other sep in the future wants to adopt our methods for other arbitrary data storage, fine, but to keep this discussion focused we need to stay on target with the task at hand which is delivering consensus on how NFTs should be stored on-chain

Author: George Kudrayvtsev (@Shaptic)
Track: Standard
Status: Draft
Created: 2021-11-22
Discussion: TODO
```

## Simple Summary
This SEP describes a specific way to store arbitrary data on-chain with the goal of standardizing the way NFTs are represented in the Stellar ledger.

## Dependencies
While not an explicit dependency, there may be a future CAP that modifies the protocol to allow a 128-byte row of ManageData to be arbitrarily divided into the string "key" and binary "data" portions, as opposed to the current model of having strict 64-byte halves.

> With that CAP, the data model introduced [below](#specification) could/should change to be friendlier and more optimized.

Reviewer 2: I think the value that this delivers would greatly simplify this SEP, and the fact the CAP would be very small both in specification and implementation makes it attractive as a quick win that'd also reduce the space NFTs consume in stellar-core instances. It'd remove the need for base91 encoding from this SEP, which while neat and the best thing we can do under the current circumstances, is also unfortunate. @Shaptic Are you planning to write a CAP for this idea?

### Representing the Data
With the encoding described [above](#encoding-binary-data), we can now discuss the actual storage. Again, we'll use the `ManageData` operation to store the binary data. 

The first entry _may_ be a [MIME type], followed by the first chunk of data, separated by a comma character (`,`). The MIME type *must* fit in a single key, and so must be <= 63 characters long.

If the MIME type is excluded, the comma _must_ still be included so that readers can easily distinguish where the type starts and the data begins.

Every subsequent entry is simply the next BasE91-encoded chunk, sized to fit into the 64-length key string, and the next 64-byte binary chunk.

If the final entry fits into the key (that is, the value is not necessary), the encoder _must_ split it into a key and a value. Specifically, since the number of bytes remaining is definitely <= 64, the key _must_ be composed of dash (`-`) characters and the value _must_ be the final chunk.

For example,

| Key | Value |
|:----|:-----:|
| text/plain,'ZP^gp@..<etc>.. | .. 64 bytes .. |
| Io1Tc%ZEYkE#^IR`0e..<etc>.. | .. 64 bytes .. |
| %yO){B&mA#_1:W8{lT..<etc>.. | .. 64 bytes .. |

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer: This is a really good point actually 🤔

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer 3: Instead of using a mandatory index prefix, could we not leave the generation of unique keys to the implementation? A different size `n` (e.g. 63 second iteration) would likely break duplicate keys and would still overall be more efficient than storing an index for every key?

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer: Using a single byte vs two would be nice but at the end of the day we still need indexing in order to reassemble the keys in the appropriate order when coming from Horizon. So unless the Horizon API changes indexing for actual reassembly is necessary. In my view we implement the SEP as Horizon stands today and if it changes in the future we can update the SEP. What we may want to include is a version key in the metadata in order to hint to the client how to interpret the data in case it changes in the future either via Horizon or CAP updates

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer 2: I don't think Horizon can, nor should we expect it to, guarantee that the key-value pairs in a JSON object are ordered. [RFC7159 Section 1](https://datatracker.ietf.org/doc/html/rfc7159#section-1) states that a JSON object is an unordered collection of name-value pairs.

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Author: A strong argument for prefixes for sure (especially in light of Tyler's other comment re: Horizon), but I would counter-argue that if your data has replicating keys, you are wasting space and should compress your data, instead. However, this is a subtlety that should at least be _addressed_ by the SEP, if not solved.

Again, though, it does lend further credence to a distinct, per-entry prefix.

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Author: (I'm avoiding making definitive statements regarding the index prefix because I'm actively trying to determine what makes Horizon return things in order, first.)

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer 3: Here is one potentially radical question.

The system is designed for optimizing storage space and costs at the expense of being intuitive design-wise (using keys to store values). I am wondering if practically, it really makes a difference for users that the storage space is 50 kb or 100 kb regardless with such high storage costs  i.e. > 400 XLM is required in both cases. Would they not likely switch to off-chain storage anyways in either scenario (likely at much lower storage requirements).

And therefore, my question is: could we not simply keep using keys for keys and values for values to keep the system consistent design-wise and have less constraints for key naming? The space would be reduced by half but does not seem realistic to even get close to such limits due to costs anyways.

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer 2: The Go stdlib orders keys using this definition:

> The map keys are sorted and used as JSON object keys by applying the following rules, subject to the UTF-8 coercion described for string values above:
>
> - keys of any string type are used directly
> - encoding.TextMarshalers are marshaled
> - integer keys are converted to strings

Ref: https://pkg.go.dev/encoding/json#Marshal

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer 2: >could we not simply keep using keys for keys and values for values to keep the system consistent design-wise and have less constraints for key naming?

💯 I think there's a lot of value in this. In the near future we (@Shaptic or I) plan to propose a CAP that will make any attempts to place data in the key pointless anyway by allowing the value field to expand to use any unused key space.

If we go with this slightly less space efficient approach, the SEP is so much simpler to implement.

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer: > And therefore, my question is: could we not simply keep using keys for keys and values for values to keep the system consistent design-wise and have less constraints for key naming?

Short answer no. I think we should support a hyper optimized standard for arbitrary data storage on-chain. I for one will opt for higher cost for on-chain storage whenever it makes sense. I don't like IPFS. A lot of the data I store is pixel art so it tends to be incredibly small already. However there is a world of difference in the Colorglyph use case when you're talking about 100 XLM vs 20 XLM when scaled to tens of thousands of glyphs. It's a difference of affordable vs not. Reliable permeant storage isn't cheap, and it's paramount we make it as cheap as possible. 

Ideally we'd get a CAP that allows for the key:value paradigm to become more flexible so this goofiness is obfuscated away and it's more a simple 0.5 XLM per 128 bytes of data calculation. Until then though I think it's worth it to push for a standard to prove why it would be wise to implement a CAP. Maybe there too we can address the indexing issue and allow for a more raw storage mechanic billed at a rate of 0.5 XLM per 128 bytes and do away with key:value all together. You pick how you want to store and surface data. Ultimately we just need to push some standard and SQ and CG have explored many various on-chain standards and some form of what I've laid out here: https://github.com/stellar/stellar-protocol/pull/1090#issuecomment-983029317 is what currently makes the most sense to me.

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer: > 💯 I think there's a lot of value in this.

In short if this SEP goes this route my vote is we abandon it. I won't use it. It'll be too expensive. I'm fine pausing SEP work until we have more clarity around CAP and/or Horizon changes but I still think there's value in formalizing a standard and using that to drive deeper improvements at more foundational layers. The point here never was to use keys and values the way they were intended to but to cram the absolute max amount of usable data into an account as a permanent data store. Sacrificing some efficiency for standards around mime-type hints and additional metadata fields is fine. Losing 64 bytes per entry entirely negates the reason we need standards in the first place and we'll revert back to everyone just smushing max data into accounts however it fits their singular isolated use case. This is first about saving money while maintaining on-chain storage, then about intended use case standards.

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Reviewer 3: > Ideally we'd get a CAP that allows for the key:value paradigm to become more flexible so this goofiness is obfuscated away and it's more a simple 0.5 XLM per 128 bytes of data calculation. Until then though I think it's worth it to push for a standard to prove why it would be wise to implement a CAP.

The CAP would be an ideal solution in the future. As I understand it would still require a SEP to define the nomenclature for keys so maybe the SEP here could also define the required key nomenclature so we can start adopting a more consistent approach with the keys optimization yet still have the flexibility of backward-compatibility for use cases like Colorglyph which require a working standard now.

> | Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |

Author: Lost this thread in the (awesome) back-and-forth we have here. Just chiming in on this:

>  I am wondering if practically, it really makes a difference for users that the storage space is 50 kb or 100 kb regardless with such high storage costs i.e. > 400 XLM is required in both cases. 

I'd say absolutely it does. A 2x size bump is the difference between Axie Infinities and Bored Apes ([article](https://medium.com/web-design-web-developer-magazine/at-what-pixel-dimensions-are-most-generative-nft-art-projects-built-9344d3998ac0), napkin math). The storage cost (around 150-200 USD at current XLM costs) is mildly prohibitive, but it also becomes part of the intrinsic value of the NFT itself.

Agreed on all counts that a CAP modifying `ManageData` would greatly simplify this SEP; perhaps in v2..

> | `-`                         | B' |

Reviewer: As mentioned in my comment earlier you can have empty values. They just need to be empty Buffers. e.g. `Buffer.alloc(0)` for example.

  1. An account is created to represent the digital asset on the ledger
  2. The account stores the entire digital asset within its managed data entries.
  3. The account issues a Stellar asset representing ownership over the digital asset we just stored.
  4. Any account holding that Stellar asset lays claim to the digital asset.

Notice that ownership of the token directly conveys ownership of a unit of the asset, entirely within the Stellar ledger. Other scenarios such as locking down the issuing account to avoid supply increases, ownership transfers via payments, etc. are built on top of this basic model and are thus extraneous to this SEP.

Note also that this SEP does not address concerns around *immutability*. While there is an optional checksum component (addressed later), the account itself is free to modify the data at-will if it is not locked.

Our exclusive focus in this SEP is Step 2: **storing the asset within the account.**

### Data Format
We begin by describing the raw data format, then dive into each component:

  - index (2 bytes)
  - version (1 byte)
  - metadata length, `n` (varying)
  - metadata (`n` bytes)
  - binary data (varying)

For example,

>     00|1|26|text/plain;l=13,image/webp|TPwJh>}A"=r@@Y?Fetc...

Reviewer: ```suggestion
    00|1|26|text/plain;s=13,image/webp|TPwJh>}A"=r@@Y?Fetc...
```

A valid SEP-39 entry always begins with `00`, indicating that this is the 0th row of data. In fact, every row's key _must_ begin with a two-character, [Base36] index.

---------

**Author's note**: This portion is definitely up for debate. Base36 (well, more accurately, a radix of 36) allows us to shave a byte off of each row (saving up to 1kB total), but we could stick to a 000-999 model (stringified integer) for simplicity. We could also do `aa` through `zz`, because as noted, the indices are _purely_ for sorting. 

Note that supporting a 36-radix is trivial in any language, since we know the value is strictly an integer. For example,

```python
import string 
ALPHABET = string.digits + string.ascii_lowercase

def int_to_b36(i: int) -> str:
    d, r = divmod(i, len(ALPHABET))
    return ALPHABET[d] + ALPHABET[r]

def b36_to_int(s: str) -> int:
    return ALPHABET.index(s[0]) * len(ALPHABET) + ALPHABET.index(s[1])
```

> 

Reviewer: Here it is in JS

> 

Author: Even better: you can use `toString` and `parseInt` (see [here](https://stackoverflow.com/questions/57803/how-to-convert-decimal-to-hexadecimal-in-javascript)):

```js
int_to_b36 = i => i.toString(36).padStart(2, '0');
b36_to_int = s => parseInt(s, 36);
```

> 

Reviewer: ```suggestion
function int_to_b36(i) {
    return i.toString(36).padStart(2, '0')
}

function b36_to_int(s) {
    return parseInt(s, 36)
}
```

---------

#### Version
The first entry's index is followed by the version string. For the current version of this SEP, this is just `1`.

#### Metadata
The **metadata** is a comma-separated list of [Media Type]s alongside any associated parameters.

> It is prefixed by a **length**, which is just a stringified integer (that is, a 32-byte metadata will have "32" as its length). When parsing, you should consume characters until hitting a non-digit, since a valid [Media Type] will never start with a digit. Note that row indices are not included in the metadata length.

Reviewer 2: >When parsing, you should consume characters until hitting a non-digit

This seems brittle to me, and it may be difficult to parse such a message safely, or at the least it might be easy to pass it in a way that is unsafe. e.g. If you read forever because someone sends you a long stream of numbers. I think we should follow in the footsteps of other message formats, which is ultimately what we're defining here, and either define a fixed number of bytes that the length fits into, or define a clear signal for how long the length should be. One way to do this is to specify a length-length, or another is to use a variable integer format where 7 bits of integer value are encoded into every byte with the high-bit indicating there's another byte, such as how protocol buffers encode integers.

I think we should lean heavily towards this being simple and easy to implement, but on the same note I see the desire to want to optimize this as much as possible. We could use XDR to define this format. That would make for increased safety, but won't necessarily be efficient since XDR pads out to multiples of 4 and all lengths are 4 bytes. 

The simpler thing, I think, would be to encode the length as binary uint16.

Reviewer: I'll defer to @Shaptic on this one but if we do this do we also need to do anything with the `l=` param? Or will the fact that it's encapsulated in metadata context keep it from key:value overflow issues? We could also just cap the length to the max possible bytes. Anything beyond that should be considered an error. So (62+64)×1,000=126,000. That would be the max integer so I think we can leave as is with the caveat for interpreters to fail if the meta length or any individual media type is > 126000.

Author: We can't use protobufs nor xdr nor int16 because this will be in the key, and thus printable ASCII @leighmcculloch. That's part of the overall challenge, unfortunately. I do agree that it's brittle, though...

I think it's reasonable to say that the length (a) _must_ fit in the first key and (b) _must_ be <= 126000. The question then becomes: should we always have 6 characters for metadata length, e.g. `000014`, or should we say "up to 6 digits"? We could also use hex, so `00000` -> `1ec30` are the valid lengths.

Thankfully, the efficiency sacrifice is not on a per-row basis, so the implementation-ease vs. efficiency trade-off is more reasonable. Thoughts?

> do we also need to do anything with the `l=` param?

I think because it's not relevant until the entire metadata + buffer is parsed, it's okay. Obviously someone can have phony lengths, but this is natural to defend against.

Reviewer: I vote leave it as is. Mixing different int parsers get's weird. Use our custom parser for the index, plain for the version, hex for the metadata length, plain again for the metadata lengths. I'd prefer to leave it as the custom for index, plain for everything else with the understanding that the metadata length (a) must fit in the first key and (b) must be <= 126000.

Author: Yep, agreed on all counts :+1:

There is only one standardized parameter:

>   * `l`: this defines the *length* (in _raw_ bytes, not encoded data) of this media type within the binary data

Reviewer: ```suggestion
  * `s`: this defines the *size* (in _raw_ bytes, not encoded data) of this media type within the binary data
```

You may include others as you see fit.

> If there is more than one [Media Type], the length parameter `l` _must_ be set for all but the last type. If there is only one, it may be omitted entirely.

Reviewer: ```suggestion
If there is more than one [Media Type], the size parameter `s` _must_ be set for all but the last type. If there is only one, it may be omitted entirely.
```

Media types are separated by a comma (`,`) character. For example, this metadata tells us that there text followed by an image within the binary buffer:

>     text/plain;l=13,image/png

Reviewer: Can we also standardize the `k=` field for denoting the `key` field for referencing the asset in question? As this is an NFT SEP it would make sense to have a standard marker mechanic for referencing assets within the binary blob.

I also think the `k=name`, `k=description` and `k=image` should be referenced as standard NFT fields. I'm assuming @FredericRezeau will include these in his off-chain NFT SEP so it would make sense to align some nomenclature here. It shouldn't be required imo but if we want wallets to display name, description and image across the board there will need to be some consensus around standard `k=` fields. If some folks use `n=` or `k=title` things will get messy.

Or should this be a separate SEP? 🤔  @leighmcculloch thoughts?

Reviewer 2: I'm a bit lost. How will an asset be simultaneously text/plain and image/png?

Reviewer: So the "asset" is actually a concatenated buffer containing 1 or more assets. So in the Colorglyph case the "asset" will be a name buffer, a description buffer and a png buffer all concatenated together. The metadata details how to split up the buffer back into its individual parts.

Author: I'm a bit lost, too, but in a different way: Can you elaborate on what a `key` is? You can already reference a specific chunk because it will follow the metadata order, e.g. the 0th chunk.

Reviewer: It's intention is for an NFT naming convention specifically to help wallets parse these sep39 accounts. I'm leaning though that those conventions should be a separate SEP leaning more heavily on @FredericRezeau for input. What values should be included in a sep39 account when being used for an NFT? It will use the `k=` metadata param but what fields are required vs optional when considering the NFT use case?

Reviewer: ```suggestion
There is only one standardized parameter:

  * `l`: this defines the *length* (in _raw_ bytes, not encoded data) of this media type within the binary data

You may include others as you see fit.

If there is more than one [Media Type], the length parameter `l` _must_ be set for all but the last type. If there is only one, it may be omitted entirely.

Media types are separated by a comma (`,`) character. For example, this metadata tells us that there text followed by an image within the binary buffer:

    text/plain;s=13,image/png
```

Reviewer: @Shaptic We still need to resolve https://github.com/stellar/stellar-protocol/pull/1090#discussion_r762303567

Reviewer: Ah nvm I see you've added `n=` as a param for the name 👍

Author: Just making sure: does `n` cover the `key` case? I don't think we need to add too many standardized fields (though we can add more as needed), because I think the off-chain convention that @FredericRezeau will spearhead will combine itself naturally with this one. An issuer will store metadata in one account (with _all_ of the details you'd expect), then have `nft.uri` (or whatever the standard field will be) point to a Stellar account, which implies said account has the asset in SEP-39 format.

Author: > When a wallet encounters an asset issued by an account with sep39 managedata attrs what should hint to it that it's an NFT?

Ah I see your line of reasoning; we avoid the need to depend on another convention. Maybe a `t=nft` as a `type` field? Or it's implied by being an image media type? Or we use the existing `t`itle? Lots of options.

Reviewer: My gut says this is a decisions for another SEP around categorizing sep39 usage. i.e. here are metadata fields for NFTs, for real estate, for identity proofs, etc. I also think this may be a good area for the toml standard to come into play, but again should likely be defined in a separate SEP.
